nexus/reconfigurator/execution/src/external_networking.rs

```diff
@@ -0,0 +1,1202 @@
+            log, "zone has unexpected IP allocated";
+            "allocated_ip" => ?existing_ip,
+        );
+        bail!("zone {zone_id} has a different IP allocated ({existing_ip:?})",);
+    }
+}
+
+// Helper function to determine whether a given NIC is already allocated to
+// a specific service zone.
+async fn is_nic_already_allocated(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    zone_kind: ZoneKind,
+    zone_id: OmicronZoneUuid,
+    nic: &NetworkInterface,
+    log: &Logger,
+) -> anyhow::Result<bool> {
+    // See the comment in is_external_ip_already_allocated().
+    if cfg!(test) && nic.ip.is_loopback() {
+        return Ok(true);
+    }
+
+    let allocated_nics = datastore
+        .service_list_network_interfaces(opctx, zone_id.into_untyped_uuid())
+        .await
+        .with_context(|| {
+            format!("failed to look up NICs for {zone_kind} {zone_id}")
+        })?;
+
+    if !allocated_nics.is_empty() {
+        // All the service zones that want NICs only expect to have a single
+        // one. Bail out here if this zone already has one or more allocated
+        // NICs but not the one we think it needs.
+        //
+        // This doesn't check the allocated NIC's subnet against our NICs,
+        // because that would require an extra DB lookup. We'll assume if
+        // these main properties are correct, the subnet is too.
+        for allocated_nic in &allocated_nics {
+            if allocated_nic.ip.ip() == nic.ip
+                && *allocated_nic.mac == nic.mac
+                && *allocated_nic.slot == nic.slot
+                && allocated_nic.primary == nic.primary
+            {
+                info!(log, "found already-allocated NIC");
+                return Ok(true);
+            }
+        }
+
+        warn!(
+            log, "zone has unexpected NICs allocated";
+            "allocated_nics" => ?allocated_nics,
+        );
+
+        bail!(
+            "zone {zone_id} already has {} non-matching NIC(s) allocated",
+            allocated_nics.len()
+        );
+    }
+
+    info!(log, "NIC allocation required for zone");
+
+    Ok(false)
+}
+
+async fn ensure_external_service_ip(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    zone_kind: ZoneKind,
+    zone_id: OmicronZoneUuid,
+    external_ip: OmicronZoneExternalIp,
+    log: &Logger,
+) -> anyhow::Result<()> {
+    // Only attempt to allocate `external_ip` if it isn't already assigned
+    // to this zone.
+    //
+    // Checking for the existing of the external IP and then creating it
+    // if not found inserts a classic TOCTOU race: what if another Nexus
+    // is running concurrently, we both check and see that the IP is not
+    // allocated, then both attempt to create it? We believe this is
+    // okay: the loser of the race (i.e., the one whose create tries to
+    // commit second) will fail to allocate the IP, which will bubble
+    // out and prevent realization of the current blueprint. That's
```

Comment:
Reasoning more to myself:

Interesting, I thought it would only happen for this specific action, not the entire blueprint. An alternative would be to check that the IP allocated was the same, and if so, log and proceed? (We should definitely fail if the IP allocated was different.)

I guess that's fine, because this is a sign that the other Nexus is ahead of us -- so we should give up and let that one lead. Seems reasonable.

Reply:
> Interesting, I thought it would only happen for this specific action, not the entire blueprint.

I think this gets at the bigger question of "what should the executor do on failure". Right now, any failure stops further blueprint realization. That's probably overly conservative, but relaxing it seems pretty tricky.

> An alternative would be to check that the IP allocated was the same, and if so, log and proceed?

True, but this feels similar to me to "should a background task retry failures" that we chatted about yesterday. The background task is already running periodically, so I think stopping execution on failures we expect to be transient is fine, because we'll try the whole task again soon anyway.